storage/google/src/server/GoogleServerStorage.ts

```diff
@@ -0,0 +1,269 @@
+    reference: ObjectReference,
+    transferType: "buffer"
+  ): Promise<Buffer>;
+
+  public download(
+    reference: ObjectReference,
+    transferType: "stream"
+  ): Promise<Readable>;
+
+  public async download(
+    reference: ObjectReference,
+    transferType: "local",
+    localPath?: string
+  ): Promise<string>;
+
+  public override async download(
+    reference: ObjectReference,
+    transferType: TransferType,
+    localPath?: string
+  ): Promise<TransferData> {
+    assertRelativeDirectory(reference.relativeDirectory);
+    if (transferType === "local") {
+      assertLocalFile(localPath);
+      await this._storage.downloadFile(reference, localPath);
+      return localPath;
+    }
+
+    const downloadBuffer = await this._storage.downloadFile(reference);
+
+    return bufferToTransferType(downloadBuffer, transferType);
+  }
+
+  public override async upload(
+    reference: ObjectReference,
+    data: TransferData,
+    metadata?: Metadata,
+    headers?: ContentHeaders
+  ): Promise<void> {
+    assertRelativeDirectory(reference.relativeDirectory);
+    if (typeof data === "string") await assertFileNotEmpty(data);
+
+    await this._storage.uploadFile(reference, data, metadata, headers);
+  }
+
+  public override async uploadInMultipleParts(
+    reference: ObjectReference,
+    data: MultipartUploadData,
+    options?: MultipartUploadOptions,
+    headers?: ContentHeaders
+  ): Promise<void> {
+    assertRelativeDirectory(reference.relativeDirectory);
+    if (typeof data === "string") await assertFileNotEmpty(data);
+
+    await this._storage.uploadFile(
+      reference,
+      data,
+      options?.metadata,
+      headers,
+      options?.partSize
+    );
+  }
+
+  public override createBaseDirectory(directory: BaseDirectory): Promise<void> {
+    return this._storageControl.createManagedFolder(directory.baseDirectory);
+  }
+
+  public override getListDirectoriesPagedIterator(
+    _maxPageSize: number
+  ): EntityPageListIterator<BaseDirectory> {
+    return new EntityPageListIterator<BaseDirectory>(async () => {
+      return this._storageControl.getManagedFoldersNextPage({
+        maxPageSize: _maxPageSize,
+      });
+    });
+  }
+
+  public override getListObjectsPagedIterator(
+    directory: BaseDirectory,
+    maxPageSize: number
+  ): EntityPageListIterator<ObjectReference> {
+    return new EntityPageListIterator<ObjectReference>(async () => {
+      return this._storage.getFilesNextPage({
+        directory: directory,
+        maxPageSize: maxPageSize,
+      });
+    });
+  }
+
+  public override list(directory: BaseDirectory): Promise<ObjectReference[]> {
+    return this.listObjects(directory);
+  }
+
+  public override async deleteBaseDirectory(
+    directory: BaseDirectory
+  ): Promise<void> {
+    for await (const objectPage of this.getListObjectsPagedIterator(
+      directory,
+      100
+    )) {
+      for (const object of objectPage) {
+        await this.deleteObject(object);
```

Comment:
If an error were to occur during this deletion, is there error handling above to recover? 

Also, this waits for each object to be deleted one at a time, it might be more efficient to  gather all of the promises into an array and await all simultaneously

Reply:
If some files failed to be deleted, you could retry it again.

We will see how the performance is in real use cases, because right now we don't have any performance metrics. Because this has to conform to the same `deleteBaseDirectory` method signature as other clients, we can't expose page size and deletion in parallel could be more taxing on the storage.